src/users/users.controller.ts

```diff
@@ -36,3 +37,45 @@ export default (usersService: UsersService) => new Elysia({prefix: '/users'})
     .get("/", () => {
         return usersService.findAll();
     })
+
+
+// Idea 4: use a class that will be injected into the app state
+// and it creates its own service (or the service can be injected 
+// into the class in the same plugin that creates the class itself)
+export class UsersController {
+    private readonly usersService: UsersService;
+    private readonly db: PostgresJsDatabase
+
+    constructor(db: PostgresJsDatabase) {
+        this.db = db;
+        this.usersService = new UsersService(this.db);
```

Comment:
This is a violation of SOLID principles: Controller shouldn't instantiate the service, it should only refer to it.
We should continue to use dependency injection and simply use

```typescript
constructor(private readonly usersService: UsersService) {}
```

This decoupling will allow us to easily mock the service and test the controller in isolation with something like

```typescript
const myTestUsersController = new UsersController(myMockUsersService);
```

Reply:
Dependency injection is nice when it makes sense, but overusing it adds complexity for no benefit. Even uncle Bob suggests in one of his talks to inject central dependencies that can be used to create the rest. I'd say this is even more maintainable because the initialization logic isn't spread around the codebase.

The mocking part is true, if we wanted to rigorously unittest everything then it'd be easier to have it injected.

At the end of the day this project won't grow to a size where either approach would hurt us, feel free to refactor it to your liking and merge.